larsim/Utils/TruthMatchUtils.cc

```diff
@@ -0,0 +1,178 @@
+        << idToEDepMap.size();
+    }
+
+    return highestContribIt;
+}
+
+} // namespace
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleID(const art::Ptr<recob::Hit> &pHit, const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+    if (0 == idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalTrueEnergy(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToEDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+        TruthMatchUtils::FillG4IDToEnergyDepositMap(idToEDepMap, pHit, rollupUnsavedIDs);
+
+    if (0==idToEDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToEDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalRecoCharge(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    IDToEDepositMap idToChargeDepMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+    {
+        const G4ID g4ID(TruthMatchUtils::TrueParticleID(pHit, rollupUnsavedIDs));
+        const EDeposit recoCharge(static_cast<EDeposit>(pHit->Integral()));
+        std::pair<IDToEDepositMap::iterator, const bool> insertPair(idToChargeDepMap.insert(IDToEDepositMap::value_type(g4ID, recoCharge)));
+        if (!(insertPair.second))
+            insertPair.first->second += recoCharge;
+    }
+
+    if (0 == idToChargeDepMap.size())
+        return kNoG4ID;
+
+    return MaxEDepElementInMap(idToChargeDepMap)->first;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+TruthMatchUtils::G4ID TruthMatchUtils::TrueParticleIDFromTotalRecoHits(const std::vector<art::Ptr<recob::Hit> >& pHits,
+    const bool rollupUnsavedIDs)
+{
+    std::map<G4ID, unsigned int> idToHitCountMap;
+    for (const art::Ptr<recob::Hit> &pHit : pHits)
+    {
+        const G4ID g4ID(TruthMatchUtils::TrueParticleID(pHit, rollupUnsavedIDs));
+        std::pair<std::map<G4ID, unsigned int>::iterator, const bool> insertPair(idToHitCountMap.insert(std::map<G4ID, unsigned int>::value_type(g4ID, 1)));
+        if (!(insertPair.second))
+            insertPair.first->second++;
```

Comment:
See above regarding more concise notation.

Reply:
I've opted to use the structured bindings as suggested throughout, but I do have some reservations about them.  I could not find a way in which I can declare some of the variables as const e.g. `auto [iterator, const inserted]`.  The scope of the vars isn't very long so its not much of a problem but it does slightly break the style...